### ISSUE TITLE
Temporary File Leak in NewtonsoftJsonInputFormatter

### ISSUE THREAD

**Issue by rwkarg** [dotnet/aspnetcore#22342]
### Describe the bug
NewtonsoftJsonInputFormatter may not close/delete temporary files created by `FileBufferingReadStream` when an error occurs.

### To Reproduce
- Create a controller with a method that takes a JSON payload as a `[FromBody]` parameter.
- Configure NewtonsoftJson for JSON serialization.
- Submit a request to the endpoint that provides > 30k of payload in the body but don't yet send a complete HTTP request
- Close the socket before completing sending the request
  - This happens for us when a proxy closes the connection but anything that closes the connection should trigger this behavior
- Observe the file system where `FileBufferingReadStream` writes temporary files (this defaults to the system Temp directory)
  - There will be an `ASPNETCORE_<guid>.tmp` file with the portion of the payload that was sent as the contents.
- The temporary file will not be deleted after the request finishes/fails
  - We've seen temporary files that are several days old before needing to restart due to 

### Further technical details
- ASP.NET Core 3.1

#### Original submission (for some reason I didn't get the bug template when creating the issue originally)

Using Newtonsoft for Json model binding, the body is buffered to a `FileBufferingReadStream` that will start buffering in memory, then spill over to a temporary file once a configured size threshold has been breached. The `FileBufferingReadStream` is created here:

https://github.com/dotnet/aspnetcore/blob/c858ccf7cf91ab836b915431bae01e96e931e7e5/src/Mvc/Mvc.NewtonsoftJson/src/NewtonsoftJsonInputFormatter.cs#L155-L160

This code, including the `DrainAsync`, is not protected by a try/finally block.

Later, a try block is created around deserialization where the `disposeReadStream` variable is checked to see if the read stream should be disposed.

https://github.com/dotnet/aspnetcore/blob/c858ccf7cf91ab836b915431bae01e96e931e7e5/src/Mvc/Mvc.NewtonsoftJson/src/NewtonsoftJsonInputFormatter.cs#L176-L190

In normal, successful operation this will dispose the `FileBufferingReadStream` and the temporary file will be deleted as expected.

Problem
=============
If something causes an error in the `DrainAsync` call (or any code before the try/finally block) then the `FileBufferingReadStream` will not be disposed and the temporary file will not be closed/deleted.

In our case, the failure trigger during `DrainAsync` was an intermediary proxy that was closing the connection due to the request being larger than the proxy's configured max request body limit, but any failure before the full body was read would trigger this.

The `FileBufferingReadStream` wraps a `FileStream` that is finalizable and looks like it should eventually close the file which would cause it to be deleted, but memory pressure for our service is such that we don't see a Gen 2 collection for days or a week so those finalizers aren't getting run. We end up running out of disk space from this client zealously retrying their failing call.

(Yes, we can mitigate this by setting a correct max body size on the proxy which is in line with the intended potentially large size of payload that we expect, but any other reason for failure in `DrainAsync` would also trigger this behavior. We also currently use Newtonsoft instead of `S.T.Json` due to some liberality in input format that we accept which cannot currently be configured in .NET Core 3.1)

We see files in the default temp directory (Linux `/tmp/ASPNETCORE_<guid>.tmp`) that do indeed contain the first part of the body payload for each request attempt. Some of these are several days old.

`XmlSerializerInputFormatter` and `XmlDataContractSerializerInputFormatter` appear to be mostly a copy of the `NewtonsoftJsonInputFormatter`

**Comment by rwkarg** [comment]
I don't see a direct way to get at either the FileBufferingReadStream or to get the specific file created to check the file system for setting up a failing unit test.

**Comment by pranavkm** [comment]
Thanks @rwkarg for the detailed explanation. We'll look in to this. 

In the meanwhile, you could consider manually buffering the request either in a middleware that executes before MVC or in a filter. `EnableBuffering` should dispose the buffer at the end of a request. MVC won't separately buffer if it sees that the request has been previously buffered.

```C#
// Filter 
public class BufferRequestFilter : Attribute, IResourceFilter
{
    public void OnResourceExecuted(ResourceExecutedContext context)
    {
        
    }

    public void OnResourceExecuting(ResourceExecutingContext context)
    {
        context.HttpContext.Request.EnableBuffering();
    }
}

[BufferRequestFilter]
public IActionResult Post([FromBody] MyModel model)
...

// Middleware
app.Use((context, next) =>
{
    context.Request.EnableBuffering();
    return next();
});
```

**Comment by davidfowl** [comment]
How big are your requests?

**Comment by danilobreda** [comment]
@davidfowl 
Sorry for my late response, i needed to turn on my metrics to have an answer:
My examples of requests:

- GET - Content-Length=8682(response) ... Integration latency: 675 ms
- GET - Content-Length=770(response)  ... Integration latency: 84 ms
- POST- Content-Length=9770(response) ... Integration latency: 1250ms
- POST - Content-Length=320(response)  ... Integration latency: 345ms

All of them have _Content-Type=application/json; charset=utf-8, Server=Kestrel, Access-Control-Allow-Origin=*_

I also check that the disc is filled in 4kb pieces. 
And only this happens when a highly processed endpoint is encountered, as in the case of the first return above 8682 bytes or 9770 bytes.
The 770bytes and 320 does not create 4kb on the disc.

I dont know right now if every request have this "pattern", if are only GET's or only POST's(with big 'body')

I also detected when the disc is with no space (it always happens) some endpoints (the heavier ones) respond with 500, and the rest works normally.

I **didnt** tried to call the .EnableBuffering(); my solution right know as it is runing on docker is that i clean the "instances" 1 time per week. But with a lot of access this week some endpoints start to have problems on the middle of the week and i restarted it manually.
This problem happens since 2.0 when I started using asp net core and docker.
